### DISCUSSION TITLE
Integrating Megaparsec with Effectful

### DISCUSSION THREAD

**Discussion by janmasrovira** [haskell-effectful/effectful#246] [Q&A]
Is it possible to define an effect that allows an instance of [`MonadParsec`](https://hackage.haskell.org/package/megaparsec-9.6.1/docs/Text-Megaparsec.html#t:MonadParsec) for effectul's [`Eff`](https://hackage.haskell.org/package/effectful-core-2.3.1.0/docs/Effectful.html#t:Eff)?

```
instance ParsingEff e s :> r => MonadParsec e s (Eff r) where
  ...
```
For my particular case, it would be enough to have `MonadParsec Void Text (Eff r)`.

So far, I've been unable to do it successfully. This post poses the same question and explains the situation quite well:
https://discourse.haskell.org/t/integrating-megaparsec-with-effectful/9954

**Comment by arybczak** [comment]
I didn't look closely, but I'm guessing that you should piggyback on ParsecT for everything. https://gist.github.com/CT075/6e0541728d9547e9ddf0b36a8ba50b56 looks generally alright, but I think it should probably use a `State` effect that keeps track of megaparsec's `State s e` and use `runParserT'` to preserve it between operations.

Something like this:

```
runParsec :: Eff (NonDet : Parsec e s : es) a -> Eff es a
runParsec = evalState initialParserState . runParsec' HANDLER . runNonDetParsec HANDLER . inject {- to hide the State -}
```

I'm not sure if it works out in the end though. Maybe it can't be written and you need to work with `ParsecT .. (Eff es)` directly.

**Reply by CT075** [reply]
Hi, I'm the author of the original linked post and snippet, I wasn't able to make this work.

This was several months ago and I don't remember what I did or did not try, but I didn't think of using a state to hold onto the state returned by `runParserT'`. Ultimately I decided that working with `ParsecT .. (Eff es)` ended up being the path of least resistance.

There's some additional subtleties in that using `NonDet` here causes a divergence the error behavior of ParsecT even if both branches fail (I believe it's something to do with which error gets "caught" and returned to the user).
